google_analytics/aggregation.py

```diff
@@ -92,58 +93,193 @@ def aggregate(df):
+        except ValueError:
+            return x
+
+    merged = pd.concat([train, test], sort=False).reset_index(drop=True)
+    merged['transactions'].fillna(0, inplace=True)
+    merged['transactionRevenue'].fillna(0.0, inplace=True)
+    # Create some features
+    merged['date'] = merged['date'].apply(lambda x: transform_date(x))
+    merged['weekday'] = merged['date'].apply(lambda x: x.weekday())
+    # Aggregate the dataset
+    merged = group_data(merged)
+    # Reduce categories on static columns
+    OHE_reduce = ['operatingSystem', 'country', 'browser']  # noqa
+    merged = reduce_categories(merged, OHE_reduce, selec_top_per, max_cat)
+    # create OHE
+    merged = one_hot_encode_categoricals(merged)
+    return merged
+
+
+def create_train_test(merged, x_train_dates=('2016-08-01', '2017-11-30'), y_train_dates=('2017-12-01', '2018-01-31'),
+                      x_test_dates=('2017-08-01', '2018-11-30'), y_test_dates=None):
+    """Splits the merged dataset on the defined time intervals.
+    """
+    def split_y(y):
+        """Split and create the Y values in the log form
+        """
+        y['target'] = y.groupby(['fullVisitorId'], as_index=False)['transactionRevenue'].sum()['transactionRevenue']
+        del y['transactionRevenue']
+        return y
+
+    def guarante_ids(x, y):
+        """To guarantee that the X and Y have the same fullVisitorId
+        """
+        names_x = x.columns.values
+        names_y = y.columns.values
+        merged = x.merge(y, on="fullVisitorId", how="inner")
+        x = merged[names_x]
+        y = merged[names_y]
+        return x, y
+
+    # Split train
+    x_train = merged[(merged["date"] >= x_train_dates[0]) & (merged["date"] <= x_train_dates[1])]
+    # Split the Ys
+    y_train = merged.loc[(merged["date"] >= y_train_dates[0]) & (merged["date"] <= y_train_dates[1]), ['transactionRevenue', 'fullVisitorId']]
+    y_train = split_y(y_train)
+    if y_test_dates is not None:
+        y_test = merged.loc[(merged["date"] >= y_test_dates[0]) & (merged["date"] <= y_test_dates[1]), ['transactionRevenue', 'fullVisitorId']]
+        y_test = split_y(y_test)
+
+    # Split test
+    x_test = merged[(merged["date"] >= x_test_dates[0]) & (merged["date"] <= x_test_dates[1])]
+
+    # create dynamic features
+    x_train = aggregate(x_train)
+    x_train.columns = [x.replace('2016', '1') for x in x_train.columns]
+    x_train.columns = [x.replace('2017', '2') for x in x_train.columns]
+
+    x_test = aggregate(x_test)
+    x_test.columns = [x.replace('2017', '1') for x in x_test.columns]
+    x_test.columns = [x.replace('2018', '2') for x in x_test.columns]
+
+    # Guarantee that the names are the same in train and test
+    names_train = set(x_train.columns.values)
+    names_test = set(x_test.columns.values)
+    names = list(names_train.intersection(names_test))
+
+    # Guarantee the same users
+    x_train, y_train = guarante_ids(x_train, y_train)
```

Comment:
We also need to drop visitors that are not buying, not visiting more than once before nov and are not visiting in nov.